source/common/router/upstream_request.h

```diff
@@ -35,6 +35,11 @@ class GenericConnPool : public Logger::Loggable<Logger::Id::router> {
 public:
   virtual ~GenericConnPool() = default;
 
+  // Initializes the connection pool. This must be called before the connection
+  // pool is valid, and can be used.
+  virtual bool initialize(Upstream::ClusterManager& cm, const RouteEntry& route_entry,
```

Comment:
I am wondering if we can avoid passing the protocol here. But the alternatives are highly coupled with cluster :)

Reply:
Based on the config proposal that @alyssawilk and I discussed today, I think that ultimately we are going to have to pass the cluster here, as that is where all of the selection logic is going to be stored that will determine which concrete impl to initialize?

I think my larger question is can we avoid an `initialize()` function entirely? It seems like what we actually will end up needing is a factory which produces a conn pool? I think that is what is going to be needed for the final impl that is pluggable? Alyssa do you want to tackle that now or later? (Personally I think it might be worth it to develop the factory in this PR and use it, even if there is no config that drives it.)

Reply:
I don't think we can avoid an initialize function or some sort of validity check.
Agree we'll need a factory.  My thought was to merge this, and then do the factory and config over in https://github.com/envoyproxy/envoy/pull/11327

I think basically that PR will get the config from the cluster, then use that to generate a get a third type of pool, the envoy.default one which creates TCP or HTTP based on "is this a connect request"

Reply:
actually I take that back - we can avoid the initialize function if we allow the factory creation function to return nullptr, if we prefer that mode

Reply:
Yeah this is what I had assumed we would do. Do you want to do that in this PR or in a follow up?